<launch>
<!--  TODO: make these parameters local -->
 <param name="robot_description" textfile="$(find knex_ros)/knex.urdf" />
 
 <node name="robot_state_publisher" pkg="robot_state_publisher" type="state_publisher" />

  <node pkg="knex_ros" type="knex_arduino_connector.py" name="knex_arduino_connector" output="screen"/>
  <node pkg="differential_drive" type="diff_tf.py" name="diff_tf" output="screen">
     <rosparam param="rate">8.0</rosparam>
     <rosparam param="base_width">0.245</rosparam>
  </node>
  <node pkg="knex_ros" type="range_filter.py" name="range_filter" output="screen">
     <rosparam param="rolling_pts">10</rosparam>
  </node>
  <node pkg="knex_ros" type="range_to_laser.py" name="range_to_laser" output="screen"/> 
  <node pkg="differential_drive" type="wheel_scaler.py" name="wheel_scaler" output="screen"/>
  <node pkg="differential_drive" type="pid_velocity.py" name="lpid_velocity" output="screen">
      <rosparam param="Kp">150</rosparam>
      <rosparam param="Ki">150</rosparam>
      <rosparam param="Kd">.000002</rosparam>
      <rosparam param="out_min">-255</rosparam>
      <rosparam param="out_max">255</rosparam>
      <rosparam param="rate">30</rosparam>
      <rosparam param="rolling_pts">5</rosparam>
      <remap from="wheel" to="lwheel"/>
      <remap from="motor_cmd" to="lmotor_cmd"/>
      <remap from="wheel_vtarget" to="lwheel_vtarget"/>
  </node> 
  <node pkg="differential_drive" type="pid_velocity.py" name="rpid_velocity" output="screen">
      <rosparam param="Kp">150</rosparam>
      <rosparam param="Ki">150</rosparam>
      <rosparam param="Kd">.000002</rosparam>
      <rosparam param="out_min">-255</rosparam>
      <rosparam param="out_max">255</rosparam>
      <rosparam param="rate">30</rosparam>
      <rosparam param="rolling_pts">5</rosparam>
      <remap from="wheel" to="rwheel"/>
      <remap from="motor_cmd" to="rmotor_cmd"/>
      <remap from="wheel_vtarget" to="rwheel_vtarget"/>
  </node> 
  
  
  <node pkg="gmapping" type="slam_gmapping" name="gmapping" output="screen">
    <remap from="scan" to="/laser"/>
    <rosparam param="odom_frame">odom</rosparam>
    <rosparam param="linearUpdate">0.04</rosparam> <!-- Process a scan each time the robot translates this far -->
    <rosparam param="angularUpdate">0.1</rosparam> <!-- Process a scan each time the robot rotates this far -->
    <rosparam param="kernelSize">0.5</rosparam>    <!-- The kernel in which to look for a correspondence -->
    <rosparam param="delta">0.05</rosparam>        <!-- resolution of the map -->
    <rosparam param="lstep">0.005</rosparam>
    <rosparam param="astep">0.05</rosparam>
    <rosparam param="iterations">10</rosparam>
    <rosparam param="lsigma">1.5</rosparam> <!-- The sigma of a beam used for likelihood -->
    <rosparam param="xmin">-10.0</rosparam>
    <rosparam param="xmax">10.0</rosparam>
    <rosparam param="ymin">-10.0</rosparam>
    <rosparam param="ymax">10.0</rosparam>
    <rosparam param="llsamplerange">0.01</rosparam>
    <rosparam param="llsamplestep">0.01</rosparam>
    <rosparam param="maxRange">1.0</rosparam>
    <rosparam param="srr">0.001</rosparam>
    <rosparam param="srt">0.002</rosparam>
    <rosparam param="str">0.001</rosparam>
    <rosparam param="stt">0.002</rosparam>
  </node>
  
  <node pkg="differential_drive" type="virtual_joystick.py" name="virtual_joystick" output="screen"/>
  <node pkg="differential_drive" type="twist_to_motors.py" name="twist_to_motors" output="screen">
    <rosparam param="base_width">0.245</rosparam>
  </node>
  
  <node pkg="rviz" type="rviz" name="rviz" output="screen"/>
  
</launch>